readme.md
[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0) 

## SMILE (Social Media Intelligence and Learning Environment)
![logo](www/public/bootstrap/img/logo/smile-logos/square3.png)

### For users or potential users:
- Watch the **[Overview video](https://www.youtube.com/watch?v=ySb-8TwJ0Ds)**
- Contact us the **[SRTI lab](https://srtilab.techservices.illinois.edu/about/)** if you have any question: <a href="mailto:srti-lab@illinois.edu">srti-lab@illinois.edu</a>
- [SMILE Playground](https://smm-tester.ncsa.illinois.edu) coming soon...

### For developers:
#### Structure of SMILE:
![Architecture](./www/docs/SMILE_architecture.png)

#### Prerequisite:
- You have **Nodejs** installed: https://nodejs.org/en/download
   * In the ```www/``` directory of SMILE, install required libraries ```npm install```
   * In the ```www/graphql/``` directory of SMILE, install required libraries ```npm install```
- Place the ```main_config.json``` file under the path ```www/``` and place the ```graphql_config.json``` file under path ```www/graphql/```
   * These two files contain credentials to acess **AWS**, **Box**, **Dropbox**, **Google drive**, **Reddit**, and **Twitter**
   * The content of ```main_config.json``` looks like this:
   >{
      "box":{
        "client_id": "#####################",
        "client_secret":"#####################"
      },
      "dropbox":{
        "client_id": "#####################",
        "client_secret":"#####################"
      },
      "google":{
        "client_id":"#####################",
        "client_secret":"#####################"
      },
      "reddit":{
        "client_id":"#####################",
        "client_secret":"#####################"
      },
      "twitter":{
        "client_id":"#####################",
        "client_secret":"#####################"
      },
      "aws":{
        "access_key":"#####################",
        "access_key_secret":"#####################"
      }
    }
   * The content of ```graphql_config.json``` looks like this:
   >{
      "twitter":{
        "client_id":"#####################",
        "client_secret":"#####################"
      }
    }

   * You have to set up **AWS** access to AWS lambda, Batch, S3 bucket, Elasticsearch, and EC2.

#### Configuration:
- Clone this repository to your local disk, ```git clone https://cwang138@opensource.ncsa.illinois.edu/bitbucket/scm/smm/smm-smile.git SMILE```. 
- Install **concurently** library ```npm install concurently -g```
   * Concurrent library is used to run SMILE sever and GraphQL server at the same time. To avoid confusion, just isntall
concurrently library globally with a ```-g```.
- Test the **analytics server** ```cd SMILE/www && npm test```
- Test the **GraphQl data server** ```cd SMILE/www/graphql && npm test```
- **RUN** concurrently ```cd SMILE/www && npm start```. The analytic tools run on http://localhost:8001, and the GraphQL data server runs on http://localhost:5050/graphql

#### Computation:
- The business logic is handled by this repository, while the computation part is deployed on AWS Lambda,
 and AWS Batch for long running process. 
You can find the deployed python code in [smm-analytics repository](https://opensource.ncsa.illinois.edu/bitbucket/projects/SMM/repos/smm-analytics/browse)
   * **lambda_classification_predict/** , **lambda_classification_split/**, and **lambda_classification_train/** contains 
   the text classification algorithms which are deployed on AWS Lambda. For larger datasets, we run the same algorithms on AWS Batch, which 
   you can find the corresponding scripts in **batch/batch_classification_predict.py**, **batch/batch_classification_split.py**, and 
   **batch/batch_classification_train.py**
   * **lambda_preprocessing/** and **batch/batch_preprocessing.py** contains the text pre-processing scripts
   * **lambda_sentiment_analysis/** and **batch/batch_sentiment_analysis.py** contains the sentiment analysis scripts
   * **lambda_network_analysis/**, and **batch/batch_network_analysis.py** contains the network analysis scripts
   * **batch/RedditComment.py** expands the top level Reddit submission and comments given a collection of threads
   * **lambda_histogram/** plot a histogram given the search results of social media
   
Please contact the **[SRTI lab](https://srtilab.techservices.illinois.edu/about/)** to request assistance on setting up the SMILE server.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.3.0] - 2023-09-14
### Fixed
- Github action to auto generate docker container [#93](https://github.com/ncsa/standalone-smm-smile/issues/93)

### Added
- Enable email sending [#72](https://github.com/ncsa/standalone-smm-smile/issues/72)

## [0.2.9] - 2023-08-04
### Added
- Environment variable to turn Clowder on and off [#86](https://github.com/ncsa/standalone-smm-smile/issues/86)
- Replace Twitter API v1 with X (Twitter) API v2 and implement OAUTH2 [#10](https://github.com/ncsa/standalone-smm-smile/issues/10)

## [0.2.8] - 2023-07-20
### Changed
- Twitter stop supporting v1.1 endpoints hence disable integration with Twitter

## [0.2.7] - 2023-06-08
### Fixed
- typo in import config file [#68](https://github.com/ncsa/standalone-smm-smile/issues/68)
- re-enable Twitter 

## [0.2.6] - 2023-06-01
### Changed
- environment variables related to minio from host_ip to minio_url [#62](https://github.com/ncsa/standalone-smm-smile/issues/62)
- temporarily disable Twitter [#65](https://github.com/ncsa/standalone-smm-smile/issues/65)

## [0.2.5] - 2022-05-11
### Changed
- disable and hide image crawling and reddit comment collections [#56](https://github.com/ncsa/standalone-smm-smile/issues/56)
- enable connection to clowder [#53](https://github.com/ncsa/standalone-smm-smile/issues/53)

### Fixed
- correct versioning for github action container building [#54](https://github.com/ncsa/standalone-smm-smile/issues/54)

## [0.2.4] - 2023-04-28
### Changed
- updated footer [#34](https://github.com/ncsa/standalone-smm-smile/issues/34)
- update citation [#33](https://github.com/ncsa/standalone-smm-smile/issues/33)
- disable hotkeys for now [#48](https://github.com/ncsa/standalone-smm-smile/issues/48)

### Added
- use import keywords to construct tag [#32](https://github.com/ncsa/standalone-smm-smile/issues/32)

### Fixed
- authorize reddit icon not clickable [#35](https://github.com/ncsa/standalone-smm-smile/issues/35)
- top bar doesn't show user account on analysis page [#31](https://github.com/ncsa/standalone-smm-smile/issues/31)
- delete data doesn't delete search tag [#28](https://github.com/ncsa/standalone-smm-smile/issues/28)
- reddit search dry run increase result number [#29](https://github.com/ncsa/standalone-smm-smile/issues/29)

## [0.2.3] - 2023-04-20
### Changed
- register new mailing list instead of srti lab [#12](https://github.com/ncsa/standalone-smm-smile/issues/12)
- add github action to automatic build and publish [#22](https://github.com/ncsa/standalone-smm-smile/issues/22)
- renamed minio related environment variable [#38](https://github.com/ncsa/standalone-smm-smile/issues/38)

## [0.2.2] - 2023-04-20
### Changed
- made Redis URL as environment variable [#16](https://github.com/ncsa/standalone-smm-smile/issues/16)
- convert s3 url to environment variable [#18](https://github.com/ncsa/standalone-smm-smile/issues/16=8)

## [1.2.8] - 2020-01-09
### Changed
- fix frontend bug on classification when click "split", "train", "predict" button

## [1.2.4] - 2019-11-07
### Changed
- use forever library to auto-restart servers when it fails
- populate more graphql errors back to SMILE server then to the browser
- fix bugs and change logic in search pagination
- update citation to the future generation computer publication

## [1.2.3] - 2019-10-25
### Changed
- UI interface change based on Hubzero usability report

### Added
- POST /dryRun endpoint to do a "pre" search instead of save the search results directly
- GET /citation endpoint to list the citation

## [1.2.2] - 2019-09-13
### Changed
- bugfix the urls schema type in twtTweetType.js from String to List

## [1.2.1] - 2019-08-20
### Changed
- Add ids to analysis output (hence rotate the aws lambda for NW, SA, PP)
- Add restrictions and hint text for upload user dataset

## [1.2.0] - 2019-07-26
### Added
- Add features to collect images from social media sources

### Changed
- Modify the Authorization page when search from social media data sources

## [1.1.9] - 2019-07-26
### Added
- Add sentiment classification using debiased word embedding algorithm

## [1.1.8] - 2019-06-24
### Added
- Add Gensim LDA topic modeling as a tool in SMILE analytics

## [1.1.7] - 2019-06-17
### Changed
- Add an identity check on the server side to check if the requested path (from frontend) 
to access a certain s3 folder matches the user's identity

## [1.1.6] - 2019-05-28
### Added
- Add SVC, Decision Tree, and Adaboost classifier alogrithm for text classification

### Changed
- Highlight the pipeline download in the popup of text classification
- Fix UID bug in text classification

## [1.1.5] - 2019-05-21
### Added
- Add Twitter timeline as a datasource in GraphQL.

### Changed
- Query interface updated.

### Removed
- Search for twitter user feature removed.



www/app.js
require('dotenv').config();
var express = require('express');

var path = require('path');
var http = require('http');
var logger = require('morgan');
var cookieParser = require('cookie-parser');
var bodyParser = require('body-parser');
var session = require('express-session');
const passport = require('passport');
const OAuth2Strategy = require('passport-oauth2').Strategy;
const redis = require('redis');
const fetch = require('node-fetch');

var lambdaRoutesTemplate = require(path.join(__dirname, 'scripts', 'helper_func', 'lambdaRoutesTemplate.js'));
var batchRoutesTemplate = require(path.join(__dirname, 'scripts', 'helper_func', 'batchRoutesTemplate.js'));
var LambdaHelper = require(path.join(__dirname, 'scripts', 'helper_func', 'lambdaHelper.js'));
var BatchHelper = require(path.join(__dirname, 'scripts', 'helper_func', 'batchHelper.js'));
var RabbitmqSender = require(path.join(__dirname, 'scripts', 'helper_func', 'rabbitmqSender.js'));
var S3Helper = require(path.join(__dirname, 'scripts', 'helper_func', 's3Helper.js'));
var fs = require('fs');
var app = express();

/**
 * default path from environment file and set it global; maybe not be used
 */
smileHomePath = path.join(process.env.HOME, 'smile');
s3FolderName = process.env.USER || 'local';
SINGLE_USER = 'true';
CLOWDER_ON = process.env.CLOWDER_ON || 'false';
CLOWDER_BASE_URL= process.env.CLOWDER_BASE_URL || "http://clowder.localhost/";
email = true;

/**
 * determine which version of deployment: dockerized vs usual
 */
if (process.env.DOCKERIZED === 'true') {
    // determine credentials either from file or from environment variable
    REDIS_URL = process.env.REDIS_URL;
    RABBITMQ_URL = process.env.RABBITMQ_URL;
    SMILE_GRAPHQL_URL = process.env.SMILE_GRAPHQL_URL;
    AWS_ACCESSKEY = process.env.AWS_ACCESSKEY;
    AWS_ACCESSKEYSECRET = process.env.AWS_ACCESSKEYSECRET;
    TWITTER_CONSUMER_KEY = process.env.TWITTER_CONSUMER_KEY;
    TWITTER_CONSUMER_SECRET = process.env.TWITTER_CONSUMER_SECRET;
    REDDIT_CLIENT_ID = process.env.REDDIT_CLIENT_ID;
    REDDIT_CLIENT_SECRET = process.env.REDDIT_CLIENT_SECRET;
    REDDIT_CALLBACK_URL = process.env.REDDIT_CALLBACK_URL;
    TWITTER_V2_CLIENT_ID = process.env.TWITTER_V2_CLIENT_ID;
    TWITTER_V2_CLIENT_SECRET = process.env.TWITTER_V2_CLIENT_SECRET;
    TWITTER_V2_CALLBACK_URL = process.env.TWITTER_V2_CALLBACK_URL;
    FLICKR_CONSUMER_KEY = process.env.FLICKR_CONSUMER_KEY;
    FLICKR_CONSUMER_SECRET = process.env.FLICKR_CONSUMER_SECRET;
    BOX_CLIENT_ID = process.env.BOX_CLIENT_ID;
    BOX_CLIENT_SECRET = process.env.BOX_CLIENT_SECRET;
    DROPBOX_CLIENT_ID = process.env.DROPBOX_CLIENT_ID;
    DROPBOX_CLIENT_SECRET = process.env.DROPBOX_CLIENT_SECRET;
    GOOGLE_CLIENT_ID = process.env.GOOGLE_CLIENT_ID;
    GOOGLE_CLIENT_SECRET = process.env.GOOGLE_CLIENT_SECRET;
    BUCKET_NAME = process.env.BUCKET_NAME;
    SINGLE_USER = process.env.SINGLE_USER;
    CLOWDER_ON = process.env.CLOWDER_ON;

    if (process.env.EMAIL_HOST === "" || process.env.EMAIL_HOST === undefined || process.env.EMAIL_HOST === null ||
        process.env.EMAIL_FROM_ADDRESS === "" || process.env.EMAIL_FROM_ADDRESS === undefined || process.env.EMAIL_FROM_ADDRESS === null) {
        email = false;
    }

    // decide what handler to use
    lambdaHandler = new RabbitmqSender();
    batchHandler = new RabbitmqSender();
    s3 = new S3Helper(true, AWS_ACCESSKEY, AWS_ACCESSKEYSECRET);

    if (SINGLE_USER === 'true') {
        global.checkIfLoggedIn = function (req, res, next) {
            req.user = {email: s3FolderName};
            return next();
        }

        global.retrieveCredentials = async function (req) {
            return new Promise((resolve, reject) => {
                if (req.session) resolve(req.session);
                else reject("There is no credential exists in the session!");
            });
        }

        global.removeCredential = async function (req, entry) {
            req.session[entry] = null;
            req.session.save();
        }

        global.setCredential = async function (req, entry, credential) {
            req.session[entry] = credential;
            req.session.save();
        };
    } else {
        // authenticate use CILogon
        passport.use(new OAuth2Strategy({
                state: true,
                authorizationURL: 'https://cilogon.org/authorize',
                tokenURL: 'https://cilogon.org/oauth2/token',
                clientID: process.env.CILOGON_CLIENT_ID,
                clientSecret: process.env.CILOGON_CLIENT_SECRET,
                callbackURL: process.env.CILOGON_CALLBACK_URL,
            }, (accessToken, refreshToken, profile, cb) => {
            fetch('https://cilogon.org/oauth2/userinfo?access_token=' + accessToken)
                .then(function (response) {
                    return response.json();
                })
                .then(function (json) {
                    process.nextTick(() => cb(null, json));
                })
            })
        );

        passport.serializeUser((user, done) => {
            done(null, user);
        });

        passport.deserializeUser(async (user, done) => {
            done(null, user);
        });

        // configure redisClient
        (async () => {
            redisClient = redis.createClient({url:REDIS_URL});
            redisClient.on('error', (err) => console.log('Redis Client Error', err));
            await redisClient.connect();
        })();

        global.checkIfLoggedIn = function (req, res, next) {
            if (req.isAuthenticated() || req.user != null) {
                return next();
            }
            res.redirect("/smile-login");
        }

        global.retrieveCredentials = async function (req) {
            return await redisClient.hGetAll(req.user.email);
        }

        global.removeCredential = async function (req, entry) {
            await redisClient.hDel(req.user.email, entry);
        }

        global.setCredential = async function (req, entry, credential) {
            await redisClient.hSet(req.user.email, entry, credential, redis.print);
            await redisClient.expire(req.user.email, 30 * 60);
        };
    }

} else {
    var config = require('./main_config.json');
    AWS_ACCESSKEY = config.aws.access_key;
    AWS_ACCESSKEYSECRET = config.aws.access_key_secret;
    TWITTER_CONSUMER_KEY = config.twitter.client_id;
    TWITTER_CONSUMER_SECRET = config.twitter.client_secret;
    REDDIT_CLIENT_ID = config.reddit.client_id;
    REDDIT_CLIENT_SECRET = config.reddit.client_secret;
    REDDIT_CALLBACK_URL = config.reddit.callback_url;
    TWITTER_V2_CLIENT_ID = config.twitterV2.client_id;
    TWITTER_V2_CLIENT_SECRET = config.twitterV2.client_secret;
    TWITTER_V2_CALLBACK_URL = config.twitterV2.callback_url;
    FLICKR_CONSUMER_KEY = config.flickr.consumer_key;
    FLICKR_CONSUMER_SECRET = config.flickr.consumer_secret;
    BOX_CLIENT_ID = config.box.client_id;
    BOX_CLIENT_SECRET = config.box.client_secret;
    DROPBOX_CLIENT_ID = config.dropbox.client_id;
    DROPBOX_CLIENT_SECRET = config.dropbox.client_secret;
    GOOGLE_CLIENT_ID = config.google.client_id;
    GOOGLE_CLIENT_SECRET = config.google.client_secret;
    SMILE_GRAPHQL_URL = "http://localhost:5050/graphql";
    BUCKET_NAME = 'macroscope-smile';
    SINGLE_USER = 'true';
    CLOWDER_ON = 'false';

    lambdaHandler = new LambdaHelper(AWS_ACCESSKEY, AWS_ACCESSKEYSECRET);
    batchHandler = new BatchHelper(AWS_ACCESSKEY, AWS_ACCESSKEYSECRET);
    s3 = new S3Helper(false, AWS_ACCESSKEY, AWS_ACCESSKEYSECRET);

    // backward compatibility; do not need multiuser capacity if deploying on macroscope
    global.checkIfLoggedIn = function (req, res, next) {
        req.user = {email: s3FolderName};
        return next();
    }

    global.retrieveCredentials = async function (req) {
        return new Promise((resolve, reject) => {
            if (req.session) resolve(req.session);
            else reject({error: "There is no credential exists in the session!"});
        });
    }

    global.removeCredential = async function (req, entry) {
        req.session[entry] = null;
        req.session.save();
    }

    global.setCredential = async function (req, entry, credential) {
        req.session[entry] = credential;
        req.session.save();
    };
}

app.use(session({
    secret: 'keyboard cat',
    resave: true,
    saveUninitialized: true,
    cookie: {maxAge: 1000 * 30 * 60}, // last half an hour?
    rolling: true
}));
app.use(express.static(path.join(__dirname, '/public')));
app.set('views', path.join(__dirname, 'views'));
app.set('view engine', 'pug');
app.use(logger('dev'));
app.use(bodyParser.json());
app.use(bodyParser.urlencoded({extended: false}));
app.use(cookieParser());
app.use(passport.initialize());
app.use(passport.session());

app.use('/', require('./routes/index.js'));

// config analytics routes
var analysesRoutesDir = path.join(__dirname, "routes", "analyses");
var analysesRoutesFiles = fs.readdirSync(analysesRoutesDir);
analysesRoutesFiles.forEach(function (route, i) {
    if (route.split(".")[0] !== ""
        && route.split(".")[1] === "json"
        && fs.lstatSync(path.join(analysesRoutesDir, route)).isFile()) {

        var routesConfig = require(path.join(analysesRoutesDir, route));

        if ("get" in routesConfig) {
            app.get("/" + routesConfig.path, checkIfLoggedIn, function (req, res) {
                var formParam = routesConfig;
                res.render('analytics/formTemplate', {
                    DOCKERIZED: process.env.DOCKERIZED === 'true',
                    title: formParam.title,
                    introduction: formParam.introduction.join(" "),
                    wiki: formParam.wiki,
                    param: formParam,
                    SINGLE_USER: SINGLE_USER==='true',
                    CLOWDER_ON: CLOWDER_ON==='true',
                    user: req.user,
                    enableEmail: email
                });
            });
        }

        if ("post" in routesConfig) {
            app.post("/" + routesConfig.path, checkIfLoggedIn, function (req, res) {
                if (req.body.selectFile !== 'Please Select...') {
                    if (req.body.aws_identifier === 'lambda') {
                        lambdaRoutesTemplate(req, routesConfig, lambdaHandler).then(data => {
                            res.send(data);
                        })
                        .catch(err => {
                            res.send({ERROR: err});
                        });
                    } else if (req.body.aws_identifier === 'batch') {
                        batchRoutesTemplate(req, routesConfig, batchHandler).then(data => {
                            res.send(data);
                        })
                        .catch(err => {
                            res.send({ERROR: err});
                        });
                    }
                } else {
                    res.end('no file selected!');
                }
            })
        }

        if ("put" in routesConfig) {

            // define an upload location
            if (!fs.existsSync(smileHomePath)) {
                fs.mkdirSync(smileHomePath);
            }
            var multer = require('multer');
            var upload = multer({dest: path.join(smileHomePath, 'uploads')});

            app.put("/" + routesConfig.path, checkIfLoggedIn, upload.single("labeled"), function (req, res) {
                s3.uploadToS3(req.file.path, req.user.email + routesConfig['result_path'] + req.body.uid
                    + '/' + req.body.labeledFilename)
                .then(url => {
                    var remoteReadPath = req.user.email + routesConfig['result_path'] + req.body.uid + '/';
                    fs.unlinkSync(req.file.path);
                    if (req.body.selectFile !== 'Please Select...') {
                        if (req.body.aws_identifier === 'lambda') {
                            lambdaRoutesTemplate(req, routesConfig, lambdaHandler).then(data => {
                                res.send(data);
                            })
                            .catch(err => {
                                res.send({ERROR: err});
                            });
                        } else if (req.body.aws_identifier === 'batch') {
                            batchRoutesTemplate(req, routesConfig, batchHandler).then(data => {
                                res.send(data);
                            })
                            .catch(err => {
                                res.send({ERROR: err});
                            });
                        }
                    } else {
                        res.end('no file selected!');
                    }
                }).catch(err => {
                    fs.unlinkSync(req.file.path);
                    res.send({ERROR: err});
                });
            });
        }
    }

});

// business logics endpoints
var busRoutesDir = path.join(__dirname, "routes", "businessLogic");
var busRoutesFiles = fs.readdirSync(busRoutesDir);
busRoutesFiles.forEach(function (route, i) {
    if (route.split(".")[0] !== ""
        && route.split(".")[1] === "js"
        && fs.lstatSync(path.join(busRoutesDir, route)).isFile()) {
        app.use('/', require('./routes/businessLogic/' + route));
    }
});

// seach endpoints
var searchRoutesDir = path.join(__dirname, "routes", "search");
var searchRoutesFiles = fs.readdirSync(searchRoutesDir);
searchRoutesFiles.forEach(function (route, i) {
    if (route.split(".")[0] !== ""
        && route.split(".")[1] === "js"
        && fs.lstatSync(path.join(searchRoutesDir, route)).isFile()) {
        app.use('/', require('./routes/search/' + route));
    }
});

// platform auth endpoints
var authRoutesDir = path.join(__dirname, "routes", "auth");
var authRoutesFiles = fs.readdirSync(authRoutesDir);
authRoutesFiles.forEach(function (route, i) {
    if (route.split(".")[0] !== ""
        && route.split(".")[1] === "js"
        && fs.lstatSync(path.join(authRoutesDir, route)).isFile()) {
        app.use('/', require('./routes/auth/' + route));
    }
});

app.get('/smile-login', passport.authenticate('oauth2', {
    scope: ['openid', 'email', 'profile']
}));

app.get('/smile-login/callback', function (req, res, next) {
    passport.authenticate('oauth2', {failureRedirect: '/smile-login'}, function (err, user, info) {
        if (err) { return res.send({ERROR: "fail to login!"}); }
        if (!user) { return res.send({ERROR: "fail to login!"}); }
        req.logIn(user, function (err) {
            if (err) { return res.send({ERROR: "fail to login!"}); }
            res.redirect('/');
        });
    })(req, res, next);
});

app.get('/smile-logout', function (req, res) {
    req.logout();
    res.redirect('/');
});

/*--------------------set server----------------------*/
var debug = require('debug');
var port = normalizePort('8001');
app.set('port', port);
var server = http.createServer(app);
server.timeout = 1000 * 60 * 10; //10 minutes

server.listen(port);
server.on('error', onError);
server.on('listening', onListening);

function normalizePort(val) {
    var port = parseInt(val, 10);

    if (isNaN(port)) {
        // named pipe
        return val;
    }

    if (port >= 0) {
        // port number
        return port;
    }

    return false;
}

function onError(error) {
    if (error.syscall !== 'listen') {
        throw error;
    }

    var bind = typeof port === 'string'
        ? 'Pipe ' + port
        : 'Port ' + port;

    // handle specific listen errors with friendly messages
    switch (error.code) {
        case 'EACCES':
            console.error(bind + ' requires elevated privileges');
            process.exit(1);
            break;
        case 'EADDRINUSE':
            console.error(bind + ' is already in use');
            process.exit(1);
            break;
        default:
            throw error;
    }
}

function onListening() {
    var addr = server.address();
    var bind = typeof addr === 'string'
        ? 'pipe ' + addr
        : 'port ' + addr.port;
    debug('Listening on ' + bind);
}
